pgml-extension/pgml_extension/utils.py

```diff
@@ -0,0 +1,47 @@
+from contextlib import contextmanager
+from datetime import datetime
+import logging
+import re
+
+_nested_timers = 0
+_previous_timer_level = 0
+_ascii_pipes = '  '
+_timer_logger = logging.getLogger('postgresml.timer')
+
+@contextmanager
+def timer(message='elapsed time:', level=logging.INFO, logger=None):
+    global _nested_timers, _previous_timer_level, _ascii_pipes, _timer_logger
+
+    if logger is None:
+        logger = _timer_logger
+
+    if level < logger.level:
+        yield
+        return
+    _nested_timers += 1
+    start = datetime.now()
+    try:
+        yield
+    finally:
+        time = datetime.now() - start
+
+        _nested_timers -= 1
+        if _nested_timers == 0:
+            _ascii_pipes = ''
+        else:
+            delta = (_nested_timers - _previous_timer_level)
+            length = _nested_timers * 2
+            if delta < 0:
+                _ascii_pipes = _ascii_pipes[0:length]
```

Comment:
Careful with these, they don't render well in logging systems that don't have a normal terminal, e.g. Datadog logs, and they make the logs hard to read.

Reply:
We should probably get smart about logging depending on the environment.